module/VuFind/src/VuFind/GeoFeatures/MapTabConfig.php

```diff
@@ -0,0 +1,140 @@
+ * @link     http://vufind.org/wiki/vufind2:recommendation_modules Wiki
+ */
+namespace VuFind\GeoFeatures;
+
+use Zend\Config\Config;
+
+/**
+ * MapTab Configuration Class
+ *
+ * @category VuFind
+ * @package  GeoFeatures
+ * @author   Leila Gonzales <lmg@agiweb.org>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:hierarchy_components Wiki
+ */
+class MapTabConfig
+{
+    /**
+     * What mapping platform should be used?
+     *
+     * @var string
+     */
+    protected $geoPlatform = null;
+
+    /**
+     * Should we display coordinates as part of labels?
+     *
+     * @var bool
+     */
+    protected $displayCoords = false;
+
+    /**
+     * Where should map labels be read from?
+     *
+     * @var string
+     */
+    protected $mapLabels = null;
+
+    /**
+     * Display graticule / map lat long grid?
+     *
+     * @var bool
+     */
+    protected $graticule = false;
+
+    /**
+     * Configuration loader
+     *
+     * @var \VuFind\Config\PluginManager
+     */
+    protected $configLoader;
+
+    /**
+     * Constructor
+     *
+     * @param \VuFind\Config\PluginManager $configLoader Configuration loader
+     */
+    public function __construct(\VuFind\Config\PluginManager $configLoader)
+    {
+        $this->configLoader = $configLoader;
+    }
+
+    /**
+     * Get the map tab configuration settings.
+     *
+     * @return array
+     */
+    public function getMapTabOptions()
+    {
+        // First check legacy location:
+        $validFields = ['displayCoords', 'mapLabels', 'graticule', 'recordMap'];
+        $options = $this->getOptions('config', 'Content', $validFields);
+        // Check geofeatures.ini [MapTab] section next:
+        if (empty($options)) {
+            $validFields = ['displayCoords', 'mapLabels', 'graticule'];
+            $options = $this->getOptions('geofeatures', 'MapTab', $validFields);
+            $geoPlatform = $this->getOptions(
+                'geofeatures', 'GeoPlatform',
+                ['geoPlatform']
+            );
+            $options['geoPlatform']=$geoPlatform['geoPlatform'];
+        }
+        if (empty($options)) {
+            // use defaults
+            $fields = ['displayCoords', 'mapLabels', 'graticule', 'geoPlatform'];
+            foreach ($fields as $field) {
+                $options[$field] = $this->$field;
+            }
+        }
+        return $options;
+    }
+
+    /**
+     * Convert a config object to an options array; return empty array if
+     * configuration is missing or incomplete.
+     *
+     * @param string $configName   Name of config file to read
+     * @param string $section      Name of section to read
+     * @param array  $validOptions List of valid fields to read
+     *
+     * @return array
+     */
+    protected function getOptions($configName, $section, $validOptions)
```

Comment:
Since MapSelectionConfig and MapTabConfig both use the same constructor and getOptions function and both have the same dependencies, this is a good candidate for inheritance. I'd recommend creating an AbstractConfig class which contains the configLoader property, constructor and getOptions method. You can then set up MapSelectionConfig and MapTabConfig as subclasses of AbstractConfig. You can then rename either MapSelectionConfigFactory or MapTabConfigFactory to AbstractConfigFactory, and change the configuration so that both classes get built by the same factory. This should accomplish exactly the same behavior with less code.

I seem to recall that your background is not in object oriented PHP, so if you need clarification on any of that, let me know and I'll be happy to elaborate -- this is a good example of why object oriented PHP can be fun, so I hope it's a useful exercise. :-)

Reply:
That's a great idea. I will give it a go and if I run into any questions I'll be sure to ask!